### DISCUSSION TITLE
Internationalization (i18n) for NiceGUI

### DISCUSSION THREAD

**Discussion by falkoschindler** [zauberzeug/nicegui#389] [Ideas / Feature Requests]
Using `ui.date` the question came up how to set the locale (#346). Apart from that there is, e.g., the connection-loss popup written in English. We might want to introduce a possibility to change the locale of a NiceGUI app and add translations for built-in messages.

**Comment by phoskee** [comment]
Hi @falkoschindler  !
Isn't i18n, but could such a thing to be useful?

![Animation](https://user-images.githubusercontent.com/125825348/234796497-f5a2a35f-35fa-4d52-aa50-be8c5fa95292.gif)

**Reply by falkoschindler** [reply]
Sure! Looks like i18n to me.

**Reply by phoskee** [reply]
Not really, it is a bit cumbersome but it might work.
Let me explain: if we load all `'language'.umd.prod.js` (or to avoid high traffic load, find a better solution) and then implement something like `ui.run(lang='it')` or `ui.run(lang='de')` etc with

```js
var lang = "{{ lang }}";  // py to js
Quasar.lang.set(Quasar.lang[lang]);  // lang choice
```

we could give the possibility to choose the language at least for quasar components, maybe!
The only step I am missing is from ui.run() to client.py

**Reply by falkoschindler** [reply]
Very interesting! The size of these language packs seems to be reasonable small: https://cdn.jsdelivr.net/npm/quasar@2/dist/lang/
From `ui.run` to `client` should work similarly to something like the `dark` parameter via globals.py.
This looks very promising. But I think we should coordinate its integration with PR #658, which is re-ogranizing all dependencies.

**Reply by phoskee** [reply]
Awesome!
You tell me how to do it, coordinate or wait for things to settle on that PR and then supplement that with another PR. At least I have time to work on it!

**Comment by phoskee** [comment]
@falkoschindler Got it!

![Animation](https://user-images.githubusercontent.com/125825348/235704570-f796a474-1ac4-4047-a243-72ff50b45a1b.gif)

For the moment, it doesn't work offline because I used this trick:

```html
<script src="https://cdn.jsdelivr.net/npm/quasar@2/dist/lang/{{lang}}.umd.prod.js"></script>`

**Reply by phoskee** [reply]
Do you think it can be implemented even if it only works online momentarily?([for this](https://github.com/zauberzeug/nicegui/discussions/389#discussioncomment-5742633))

**Reply by falkoschindler** [reply]
Yes, I think so. Adding the language packs shouldn't be a problem.

**Comment by falkoschindler** [comment]
In NiceGUI 1.2.12 we released @phoskee's PR https://github.com/zauberzeug/nicegui/pull/874. Therefore I'll close this discussion.
